### PR TITLE
Fixed NeptuneLogger support across different versions of neptune package

### LINKED ISSUE #18555: cannot find neptune.new.utils with `Neptune-client==0.16.3`
### Bug description

Pytorch Lightning 2.0.8 - `import pytorch_lightning as pl` fails with error message `No module named 'neptune.new.utils'`, but I have Neptune-client 0.16.3. More interestingly, the code runs perfectly on an env that does not even have neptune-client installed.
I have reproduced on a clean environment (see steps below), and I was able to import Lightning up until the point where I installed Neptune-client. Then, the import fails with the error message presented here.
My problem is that I need to use Lightning in an env a customer provided us, which uses Neptune-client as indicated, and I am therefore unsure on whether we can remove/change that package.
This is on Python 3.10.12.

Error traceback:
```
ModuleNotFoundError                       Traceback (most recent call last)
File [main.py] in line 7
      1 #%%
      2 # basic libraries
      3 #import  os
      4 #import  yaml
      5 
      6 # pytorch lightning
----> 7 import  pytorch_lightning           as      pl
      8 from    lightning.pytorch           import  Trainer
      9 from    lightning.pytorch.loggers   import  TensorBoardLogger

File [~/miniforge3/envs/test_env/lib/python3.10/site-packages/pytorch_lightning/__init__.py:26]
     23     _logger.propagate = False
     25 from lightning_fabric.utilities.seed import seed_everything  # noqa: E402
---> 26 from pytorch_lightning.callbacks import Callback  # noqa: E402
     27 from pytorch_lightning.core import LightningDataModule, LightningModule  # noqa: E402
     28 from pytorch_lightning.trainer import Trainer  # noqa: E402

File [~/miniforge3/envs/test_env/lib/python3.10/site-packages/pytorch_lightning/callbacks/__init__.py:29]
     27 from pytorch_lightning.callbacks.prediction_writer import BasePredictionWriter
     28 from pytorch_lightning.callbacks.progress import ProgressBar, RichProgressBar, TQDMProgressBar
---> 29 from pytorch_lightning.callbacks.pruning import ModelPruning
     30 from pytorch_lightning.callbacks.rich_model_summary import RichModelSummary
...
     56 else:
     57     # needed for tests, mocks and function signatures
     58     neptune, Run, Handler, File, stringify_unsupported = None, None, None, None, None

ModuleNotFoundError: No module named 'neptune.new.utils'
```

### What version are you seeing the problem on?

v2.0

### How to reproduce the bug

Create new environment
```
conda create -n test_env
```

Install Pytorch 1.11.0
```
conda install pytorch==1.11.0 torchvision==0.12.0 torchaudio==0.11.0 -c pytorch
```

Install Pytorch Geometric (the one for Pytorch 1.11)
```
pip install torch-scatter torch-sparse torch-cluster torch-spline-conv torch-geometric -f https://data.pyg.org/whl/torch-1.11.0+cpu.html
```

Install Pytorch Lightning (note that lightning 2.0 is compatible with Pytorch 1.11, see compatibility matrix https://lightning.ai/docs/pytorch/stable/versioning.html)
```
python -m pip install lightning
```

Install neptune-client 0.16.3
```
pip install neptune-client==0.16.3
```

Try importing Pytorch Lightning
```
import pytorch_lightning as pl
```

Obtain error message above.
```



### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version: 2.0.8
#- PyTorch Version: 1.11.0
#- Python version: 3.10.12
#- OS: McOS Ventura 13.4.1
```

</details>


cc @Blaizzy

### PR DESCRIPTION
**Pull request by Raalsky**
## What does this PR do?

Fixes #18555

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚  Documentation preview 📚 : https://pytorch-lightning--19131.org.readthedocs.build/en/19131/

### CODE REVIEW

File: src/lightning/pytorch/loggers/neptune.py [hunk 1]
```diff
@@ -39,8 +39,8 @@
 log = logging.getLogger(__name__)
 
 # neptune is available with two names on PyPI : `neptune` and `neptune-client`
-_NEPTUNE_AVAILABLE = RequirementCache("neptune>=1.0")
-_NEPTUNE_CLIENT_AVAILABLE = RequirementCache("neptune-client")
+_NEPTUNE_AVAILABLE = bool(RequirementCache("neptune>=1.0")) or bool(RequirementCache("neptune-client>=1.0"))
+_LEGACY_NEPTUNE_CLIENT = bool(RequirementCache("neptune-client<1.0"))
```
Comment (awaelchli) [inline src/lightning/pytorch/loggers/neptune.py:43]:
```suggestion
_LEGACY_NEPTUNE_CLIENT_AVAILABLE = bool(RequirementCache("neptune-client<1.0"))
```
to follow the naming pattern

File: src/lightning/pytorch/loggers/neptune.py [hunk 2]
```diff
@@ -223,8 +223,8 @@ def __init__(
         prefix: str = "training",
         **neptune_run_kwargs: Any,
     ):
-        if not _NEPTUNE_AVAILABLE and not _NEPTUNE_CLIENT_AVAILABLE:
-            raise ModuleNotFoundError(str(_NEPTUNE_AVAILABLE))
+        if not _NEPTUNE_AVAILABLE and not _LEGACY_NEPTUNE_CLIENT:
+            raise ModuleNotFoundError("`neptune` package is required. Run `pip install neptune`")
```
Comment (awaelchli) [inline src/lightning/pytorch/loggers/neptune.py:227]: The previous code was already printing this hint and additional info when the package was not found.

**Review by awaelchli — APPROVED** [review]

File: src/lightning/pytorch/loggers/neptune.py [hunk 3]
```diff
@@ -39,8 +39,8 @@
 log = logging.getLogger(__name__)
 
 # neptune is available with two names on PyPI : `neptune` and `neptune-client`
-_NEPTUNE_AVAILABLE = RequirementCache("neptune>=1.0")
-_NEPTUNE_CLIENT_AVAILABLE = RequirementCache("neptune-client")
+_NEPTUNE_AVAILABLE = bool(RequirementCache("neptune>=1.0")) or bool(RequirementCache("neptune-client>=1.0"))
+_LEGACY_NEPTUNE_CLIENT_AVAILABLE = bool(RequirementCache("neptune-client<1.0"))
```
Comment (carmocca) [inline src/lightning/pytorch/loggers/neptune.py:43]: These shouldn't be converted to bools. This class overrides __str__ providing a helpful error message to the users. See for instance the use in L227

Comment (Borda) [reply src/lightning/pytorch/loggers/neptune.py:43]:
```suggestion
_NEPTUNE_AVAILABLE = RequirementCache("neptune>=1.0") or RequirementCache("neptune-client>=1.0")
_LEGACY_NEPTUNE_CLIENT_AVAILABLE = RequirementCache("neptune-client<1.0")
```

**Review by Borda — APPROVED** [review]

File: src/lightning/pytorch/loggers/neptune.py (hunk 3 above)
Comment (carmocca) [reply src/lightning/pytorch/loggers/neptune.py:43]: @Raalsky What's the issue with having two separate variables each pointing to each class? (as in the current version)

Comment (Raalsky) [reply src/lightning/pytorch/loggers/neptune.py:43]: > These shouldn't be converted to bools. This class overrides str providing a helpful error message to the users. See for instance the use in L227

I've been trying during the development and I remember that this leads to two separate messages about the lack of one of the packages. The expected behavior is to raise/warn whenever none of `neptune-client`/ `neptune` was installed. This should be covered in: https://github.com/Lightning-AI/pytorch-lightning/pull/19131/files#diff-903c2ecc76f95d71b3f577baf15c47f0e629f6e2a4b201a344177e7908cfccd9R323

> What's the issue with having two separate variables each pointing to each class? (as in the current version)

`neptune` was introduced as a name transition of `neptune-client` and the long-term target is to get rid of `Neptune-client` package completely someday. It was introduced as a part of breaking-changes with a release of `neptune-client==1.0`. `neptune-client>=1.0` is just an alias of `neptune` package and have some breaking-changes in compare to `neptune-client<1.0.0`. We would like to still support all of the possible environments and package versions. The earlier approach was just "wrong" and had some misunderstandings of how the relationships between each of the packages looked like.

Comment (carmocca) [reply src/lightning/pytorch/loggers/neptune.py:43]: If that's the motivation, then this should work (see other suggestions too=:

```python
_NEPTUNE_AVAILABLE = RequirementCache("neptune>=1.0")
# neptune-client will be gone someday. additionally, version 1.0 had breaking changes to account for
_NEPTUNE_CLIENT_AVAILABLE = RequirementCache("neptune-client>=1.0")
_LEGACY_NEPTUNE_CLIENT_AVAILABLE = RequirementCache("neptune-client<1.0")
_NEPTUNE_GREATER_EQUAL_1_0 = bool(_NEPTUNE_AVAILABLE or _NEPTUNE_CLIENT_AVAILABLE)
```

File: src/lightning/pytorch/loggers/neptune.py [hunk 4]
```diff
@@ -335,12 +338,12 @@ def __getstate__(self) -> Dict[str, Any]:
 
     def __setstate__(self, state: Dict[str, Any]) -> None:
         if _NEPTUNE_AVAILABLE:
```
Comment (Borda) [inline src/lightning/pytorch/loggers/neptune.py:340]: this seems strange import Neptune even Neptune is not available...

File: src/lightning/pytorch/loggers/neptune.py [hunk 5]
```diff
@@ -376,12 +379,12 @@ def training_step(self, batch, batch_idx):
     @rank_zero_experiment
     def run(self) -> "Run":
         if _NEPTUNE_AVAILABLE:
```
Comment (Borda) [inline src/lightning/pytorch/loggers/neptune.py:381]: Please use a better, more desriptive name, as it is not about Neptune but version...

**Review by Borda — CHANGES_REQUESTED** [review]

**Comment by Borda** [comment]
let's update or rebase after #19265 lands :flags:

**Comment by Raalsky** [comment]
Closing in favor of https://github.com/Lightning-AI/pytorch-lightning/pull/19265

**Pull request closed without merging**